### PR TITLE
Improve numerical stability of SoftmaxWithCrossEntropy

### PR DESCRIPTION
**Pull request by cjld**
TolerableValue is always used after log. So the kLogMin is a more appropriate value than kApproInf

### CODE REVIEW

File: paddle/fluid/operators/math/cross_entropy.h [hunk 1]
```diff
@@ -26,9 +27,12 @@ struct TolerableValue {
   HOSTDEVICE T operator()(const T& x) const {
     PADDLE_ASSERT(std::is_floating_point<T>::value);
     const T kApproInf = 1e20;
+    const T kLogMin = log(std::numeric_limits<T>::min());
 
     if (x == INFINITY) return kApproInf;
-    if (x == -INFINITY) return -kApproInf;
+    // TolerableValue is always used after log. So the kLogMin is a more
+    // appropriate value than kApproInf
+    if (x == -INFINITY) return -kLogMin;
```
Comment (qingqing01) [inline paddle/fluid/operators/math/cross_entropy.h]:
```c++
std::log(std::numeric_limits<float>::min()) = - 87.3365
std::log(std::numeric_limits<double>::min()) = -708.396
```

Here, change from -1e20 to +87.3365 or +708.396

It seems wrong ?

@panyx0718 @kuke How do you think?

Comment (kuke) [reply paddle/fluid/operators/math/cross_entropy.h]: The final change LGTM

Comment (cjld) [reply paddle/fluid/operators/math/cross_entropy.h]: forgot to remove the minus...
fixed

File: paddle/fluid/operators/math/cross_entropy.h [hunk 2]
```diff
@@ -26,9 +27,12 @@ struct TolerableValue {
   HOSTDEVICE T operator()(const T& x) const {
```
Comment (panyx0718) [inline paddle/fluid/operators/math/cross_entropy.h:27]: Is TolerableValue really a good idea?

Comment (cjld) [reply paddle/fluid/operators/math/cross_entropy.h:27]: No, actually. We can calculate softmax and cross entropy at the same time, rather than calculate separately.
But It may require lots of changes. So this PR is just a temporary fix of wrong losses.

Comment (panyx0718) [reply paddle/fluid/operators/math/cross_entropy.h:27]: @qingqing01 can make this explicitly configurable by the user. So we don't do this secretly.

Comment (panyx0718) [reply paddle/fluid/operators/math/cross_entropy.h:27]: or we can calculate them at the same time

**Comment by cjld** [comment]
image

this pr is needed, this figuration shows nan occur(up), and this pr can fix(down)
